hcloud/action_waiter.go

```diff
@@ -0,0 +1,102 @@
+package hcloud
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/samber/lo"
+)
+
+type ActionWaiter interface {
+	WaitForActionsFunc(ctx context.Context, actions []*Action, handleUpdate func(update *Action)) ([]*Action, error)
+	WaitForActions(ctx context.Context, actions []*Action) ([]*Action, error)
+	WaitForAction(ctx context.Context, action *Action) (*Action, error)
+}
+
+var _ ActionWaiter = (*ActionClient)(nil)
+
+// WaitForActionsFunc implements ActionWaiter.
+func (c *ActionClient) WaitForActionsFunc(ctx context.Context, actions []*Action, handleUpdate func(update *Action)) ([]*Action, error) {
+	result := make([]*Action, 0, len(actions))
+
+	running := make(map[int64]struct{}, len(actions))
+	for _, action := range actions {
+		if action.Status == ActionStatusRunning {
+			running[action.ID] = struct{}{}
+		} else {
+			result = append(result, action)
+		}
+	}
+
+	retries := 0
+	for {
+		if len(running) == 0 {
+			break
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(c.action.client.pollBackoffFunc(retries)):
+			retries++
+		}
+
+		opts := ActionListOpts{Sort: []string{"status", "id"}}
+		for actionID := range running {
+			opts.ID = append(opts.ID, actionID)
+		}
+
+		updates, err := c.AllWithOpts(ctx, opts)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
> The early return after AllWithOpts would cause issues if the network is wonky. Ie. if you try to create a server from snapshot on WifiOnICE
> 
> Perhaps thats also just fine and should be handled outside of this method, related https://github.com/hetznercloud/hcloud-go/issues/311

I believe the actions waiter is not responsible for handling any network flakiness or implement any retry logic. This is better handled a few layers down, probably by wrapping the http client `Do` method.